cmd/hashers/argon2/calibrate.go

```diff
@@ -0,0 +1,260 @@
+package argon2
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/c2h5oh/datasize"
+	"github.com/fatih/color"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+
+	"github.com/ory/kratos/driver/configuration"
+	"github.com/ory/kratos/hash"
+	"github.com/ory/x/cmdx"
+	"github.com/ory/x/flagx"
+)
+
+type (
+	argon2Config struct {
+		c configuration.HasherArgon2Config
+	}
+)
+
+func (c *argon2Config) HasherArgon2() *configuration.HasherArgon2Config {
+	return &c.c
+}
+
+func (c *argon2Config) getMemFormat() string {
+	return (datasize.ByteSize(c.c.Memory) * datasize.KB).HumanReadable()
+}
+
+const (
+	FlagStartMemory     = "start-memory"
+	FlagMaxMemory       = "max-memory"
+	FlagAdjustMemory    = "adjust-memory-by"
+	FlagStartIterations = "start-iterations"
+	FlagParallelism     = "parallelism"
+	FlagSaltLength      = "salt-length"
+	FlagKeyLength       = "key-length"
+
+	FlagVerbose = "verbose"
+	FlagRuns    = "probe-runs"
+)
+
+var resultColor = color.New(color.FgGreen)
+
+func newCalibrateCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "calibrate [<desired-duration>]",
+		Args:  cobra.ExactArgs(1),
+		Short: "Calibrate the values for Argon2 so the hashing operation takes the desired time.",
+		Long: `Calibrate the configuration values for Argon2 by probing the execution time.
+Note that the values depend on the machine you run the hashing on.
+When choosing the desired time, UX is in conflict with security. Security should really win out here, therefore we recommend 1s.
+`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			desiredDuration, err := time.ParseDuration(args[0])
+			if err != nil {
+				return err
+			}
+			verbose := flagx.MustGetBool(cmd, FlagVerbose)
+			runs := flagx.MustGetInt(cmd, FlagRuns)
+
+			config, err := configFromFlags(cmd)
+			if err != nil {
+				return err
+			}
+			hasher := hash.NewHasherArgon2(config)
+
+			var maxMemory, adjustMemory datasize.ByteSize
+			if err := maxMemory.UnmarshalText([]byte(flagx.MustGetString(cmd, FlagMaxMemory))); err != nil {
+				return err
+			}
+			if err := adjustMemory.UnmarshalText([]byte(flagx.MustGetString(cmd, FlagAdjustMemory))); err != nil {
+				return err
+			}
```

Comment:
See below - reusing `flagx.GetByteSize` could reduce the visual clutter here a bit